src/lib/reasoners/bitv.ml

```diff
@@ -192,137 +192,144 @@ module Shostak(X : ALIEN) = struct
       let compare = compare_solver_simple_term
     end)
 
-  module Canonizer = struct
-
-    type term_aux  =
-      | I_Cte of bool
-      | I_Other of X.r
-      | I_Ext of term * int * int
-      | I_Comp of term * term
-
-    and term = term_aux alpha_term
-
-    let rec compare_term_aux t1 t2 = match t1, t2 with
-      | I_Cte true, I_Cte true
-      | I_Cte false, I_Cte false -> 0
-      | I_Cte true, I_Cte _ -> 1
-      | I_Cte _, I_Cte _ -> -1
-
-      | I_Other xterm1, I_Other xterm2 -> X.str_cmp xterm1 xterm2
-
-      | I_Ext (t, i, j), I_Ext (t', i', j') ->
-        if i <> i' then i - i'
-        else if j <> j' then j - j'
-        else compare_term t t'
-
-      | I_Comp (t1, t2), I_Comp (t1', t2') ->
-        let cmp = compare_term t1 t1' in
-        if cmp = 0 then compare_term t2 t2' else cmp
-
-      | I_Cte _, (I_Other _ | I_Ext _ | I_Comp _ ) -> 1
-      | I_Other _, (I_Ext _ | I_Comp _ ) -> 1
-      | I_Ext _, I_Comp _ -> 1
-
-      | I_Comp _, (I_Cte _ | I_Other _ | I_Ext _) -> -1
-      | I_Ext _, (I_Cte _ | I_Other _) -> -1
-      | I_Other _, I_Cte _ -> -1
-
-    and compare_term t1 t2 = compare_alpha_term compare_term_aux t1 t2
-
-    (** **)
-    let rec alpha t = match t.bv with
-      |I_Cte _ -> [t]
-      |I_Other _ -> [t]
-      |I_Comp (t1,t2) -> (alpha t1)@(alpha t2)
-      |I_Ext(t',i,j) ->
-        begin
-          match t'.bv with
-          |I_Cte _ -> [{t' with sz = j-i+1}]
-          |I_Other _ -> [t]
-          |I_Ext(t'',k,_) ->
-            alpha {t with bv = I_Ext(t'',i+k,j+k)}
-
-          |I_Comp(_,v) when j < v.sz ->
-            alpha{t with bv =I_Ext(v,i,j)}
-
-          |I_Comp(u,v) when i >= v.sz ->
-            alpha{t with bv=I_Ext(u,i-v.sz,j-v.sz)}
-
-          |I_Comp(u,v) ->
-            (alpha {sz = j-v.sz+1 ; bv = I_Ext(u,0,j-v.sz)})
-            @(alpha{sz = v.sz-i ; bv = I_Ext(v,i,v.sz-1)})
+  module Canon = struct
+    type 'a view_descr =
+      | Vcte of string
+      | Vother of 'a
+      | Vextract of 'a * int * int
+      | Vconcat of 'a * 'a
+
+    and 'a view = { descr : 'a view_descr ; size : int }
+
+    let view t =
+      match E.term_view t with
+      | { f = Bitv s; ty = Tbitv size; _ } -> { descr = Vcte s; size }
+      | { f = Op Concat; xs = [ t1; t2 ]; ty = Tbitv size; _ } ->
+        { descr = Vconcat (t1, t2); size }
+      | { f = Op Extract (i, j); xs = [ t' ]; ty = Tbitv size; _ } ->
+        assert (size = j - i + 1);
+        { descr = Vextract (t', i, j); size }
+      | { ty = Tbitv size; _ } ->
+        { descr = Vother t; size }
+      | _ -> assert false
+
+    let run ctx f = f ctx
+    let (let*) x f ctx =
+      let x, ctx = x ctx in
+      f x ctx
+    let ret x ctx = x, ctx
+    let (and*) x y =
+      let* x = x in
+      let* y = y in
+      ret (x, y)
+    let (let+) x f ctx =
+      let x, ctx = x ctx in
+      f x, ctx
+    let (and+) = (and*)
```

Comment:
A bit weird to use `x` as both a function and the result of a function application, the argument `x` could be called `g` for example.

Reply:
I think it is relatively idiomatic to have `x` be both the "stateful" `'a m = 'st -> 'a * 'st` and the "inner" `'a`, this is what happens e.g. when we do `let* x = x in ...` as well.

I'm happy to call it `mx` here, and will add a comment explaining that `(let*)` and `(let+)` implement `bind` and `fmap` for the state monad as well.